core/src/main/java/org/infinispan/commands/remote/ClusteredGetAllCommand.java

```diff
@@ -93,29 +94,25 @@ public void init(InvocationContextFactory icf, CommandsFactory commandsFactory,
       InvocationContext invocationContext = icf.createRemoteInvocationContextForCommand(command, getOrigin());
       CompletableFuture<Object> future = invoker.invokeAsync(invocationContext, command);
       return future.thenApply(rv -> {
-         Map<K, CacheEntry<K, V>> map = (Map<K, CacheEntry<K, V>>) rv;
-         if (trace) log.trace("Found: " + map);
-
-         if (map == null) {
-            return null;
+         if (trace) log.trace("Found: " + rv);
+         if (rv == null || rv instanceof Response) {
+            return rv;
          }
 
-         List<InternalCacheValue<V>> values = new ArrayList<>(keys.size());
+         Map<K, CacheEntry<K, V>> map = (Map<K, CacheEntry<K, V>>) rv;
+         InternalCacheValue<V>[] values = new InternalCacheValue[keys.size()];
+         int i = 0;
          for (Object key : keys) {
-            if (map.containsKey(key)) {
-               CacheEntry<K, V> entry = map.get(key);
-               InternalCacheValue<V> value;
-               if (entry instanceof InternalCacheEntry) {
-                  value = ((InternalCacheEntry<K, V>) entry).toInternalCacheValue();
-               } else if (entry != null) {
-                  value = entryFactory.createValue(entry);
-               } else {
-                  value = new ImmortalCacheValue(null);
-               }
-               values.add(value);
+            CacheEntry<K, V> entry = map.get(key);
+            InternalCacheValue<V> value;
+            if (entry == null) {
+               value = null;
```

Comment:
Because we're returning `InternalCacheValue[]` in the same order as we got `List<?> keys`.